### PR TITLE
Implement virtual symbols and WiX Standard Library

### PR DESCRIPTION
**Pull request by robmen**
Closes wixtoolset/issues#7913
Closes wixtoolset/issues#7914

### CODE REVIEW

File: src/wix/WixToolset.Core/Compiler_Package.cs [hunk 1]
```diff
@@ -2521,7 +2523,8 @@ private void ParseSequenceElement(XElement node, SequenceTable sequenceTable)
                         case "Dialog":
                             if (showDialog)
                             {
-                                actionName = this.Core.GetAttributeIdentifierValue(childSourceLineNumbers, attrib);
+                                actionIdentifier = this.Core.GetAttributeIdentifier(childSourceLineNumbers, attrib);
+                                actionName = actionIdentifier.Id; //this.Core.GetAttributeIdentifierValue(childSourceLineNumbers, attrib);
```
Comment (barnson) [inline src/wix/WixToolset.Core/Compiler_Package.cs:2529]: nit: remove comment

File: src/wix/WixToolset.Core/ExtensibilityServices/ParseHelper.cs [hunk 2]
```diff
@@ -423,13 +423,27 @@ public Identifier GetAttributeIdentifier(SourceLineNumber sourceLineNumbers, XAt
...
+                        break;
+
                     default:
                         return null;
                 }
 
                 value = value.Substring(separator + 1).Trim();
             }
 
+            //////if (!allowVirtual && (access == AccessModifier.Virtual || access == AccessModifier.Override))
```
Comment (barnson) [inline src/wix/WixToolset.Core/ExtensibilityServices/ParseHelper.cs:441]: clean up

File: src/wix/WixToolset.Core/Link/AddRequiredStandardDirectories.cs [hunk 3]
```diff
@@ -43,12 +43,17 @@ public void Execute()
...
                 {
-                    if (directory.Id.Id != "TARGETDIR")
-                    {
-                        directory.ParentDirectoryRef = "TARGETDIR";
-                    }
+                    parentDirectoryId = this.GetStandardDirectoryParent(directory.Id.Id, platform);
+
+                    directory.ParentDirectoryRef = parentDirectoryId;
+
+                    //if (directory.Id.Id != "TARGETDIR")
```
Comment (barnson) [inline src/wix/WixToolset.Core/Link/AddRequiredStandardDirectories.cs]: etc

File: src/wix/WixToolset.Core/Link/FindEntrySectionAndLoadSymbolsCommand.cs [hunk 4]
```diff
@@ -84,15 +92,34 @@ public void Execute()
...
+                            existingSymbol.OverrideVirtualSymbol(symbolWithSection);
+
+                            overriddenSymbols.Add(symbolWithSection.Symbol);
+                        }
                         // If the duplicate symbols are both private directories, there is a chance that they
-                        // point to identical symbols. Identical directory symbols should be treated as redundant.
-                        // and not cause conflicts.
-                        if (AccessModifier.Section == existingSymbol.Access && AccessModifier.Section == symbolWithSection.Access &&
-                            SymbolDefinitionType.Directory == existingSymbol.Symbol.Definition.Type && existingSymbol.Symbol.IsIdentical(symbolWithSection.Symbol))
+                        // point to identical symbols. Identical directory symbols are redundant and will not cause
```
Comment (barnson) [inline src/wix/WixToolset.Core/Link/FindEntrySectionAndLoadSymbolsCommand.cs:119]: Confidence is good. :)

File: src/wix/WixToolset.Core/Link/ResolveReferencesCommand.cs [hunk 5]
```diff
@@ -133,13 +133,13 @@ private List<SymbolWithSection> DetermineAccessibleSymbols(IntermediateSection r
...
             foreach (var dupe in symbolWithSection.PossiblyConflicts)
             {
-                // don't count overridable WixActionSymbols
-                var symbolAction = symbolWithSection.Symbol as WixActionSymbol;
-                var dupeAction = dupe.Symbol as WixActionSymbol;
-                if (symbolAction?.Overridable != dupeAction?.Overridable)
-                {
-                    continue;
-                }
+                //// don't count overridable WixActionSymbols
```
Comment (barnson) [inline src/wix/WixToolset.Core/Link/ResolveReferencesCommand.cs]: clean up

**Comment by robmen** [comment]
@barnson please, please, please review my error messages. I struggled so much trying to generate useful English.

File: src/wix/WixToolset.Core/Linker.cs [hunk 6]
```diff
@@ -119,8 +122,8 @@ public Intermediate Link(ILinkContext context)
                     }
                 }
 
-                // Add the missing standard action and directory symbols.
-                this.LoadStandardSymbols(find.SymbolsByName);
+                ////// Add the missing standard action and directory symbols.
```
Comment (barnson) [inline src/wix/WixToolset.Core/Linker.cs]: clean up (unless it's one cleaned up in a later commit)

**Comment by barnson** [comment]
:shipit: Minor language tweaks.

File: src/wix/WixToolset.Core/Link/ResolveReferencesCommand.cs (hunk 5 above)
Comment (robmen) [reply src/wix/WixToolset.Core/Link/ResolveReferencesCommand.cs]: Yep, cleaned up in subsequent commit.

File: src/wix/WixToolset.Core/Linker.cs (hunk 6 above)
Comment (robmen) [reply src/wix/WixToolset.Core/Linker.cs]: Yep. Cleaned up in subsequent commit.

File: src/wix/WixToolset.Core/Link/AddRequiredStandardDirectories.cs (hunk 3 above)
Comment (robmen) [reply src/wix/WixToolset.Core/Link/AddRequiredStandardDirectories.cs]: This was all removed in a subsequent commit.

File: src/wix/WixToolset.Core/Compiler_Package.cs (hunk 1 above)
Comment (robmen) [reply src/wix/WixToolset.Core/Compiler_Package.cs:2529]: Fixed locally. I will push a rebase at the end of your review @barnson.

File: src/wix/WixToolset.Core/ExtensibilityServices/ParseHelper.cs (hunk 2 above)
Comment (robmen) [reply src/wix/WixToolset.Core/ExtensibilityServices/ParseHelper.cs:441]: Fixed locally. I will push a rebase at the end of your review @barnson.

**Comment by robmen** [comment]
@barnson does the :shipit: mean overall this is good to go (with the deleting of dead code)? Or are you still reviewing?

**Comment by barnson** [comment]
GtG with messages.

**Pull request merged**